walkeeper/src/wal_service.rs

```diff
@@ -58,14 +57,14 @@ fn handle_socket(mut socket: TcpStream, conf: WalAcceptorConf) -> Result<()> {
 }
 
 /// Fetch the first 4 bytes from the network (big endian), without consuming them.
-///
 /// This is used to help determine what protocol the peer is using.
 fn peek_u32(stream: &mut TcpStream) -> Result<u32> {
     let mut buf = [0u8; 4];
     loop {
-        let num_bytes = stream.peek(&mut buf)?;
-        if num_bytes == 4 {
-            return Ok(u32::from_be_bytes(buf));
+        match stream.peek(&mut buf)? {
+            0 => anyhow::bail!("unexpected end of stream"),
+            4 => return Ok(u32::from_be_bytes(buf)),
+            _ => continue,
```

Comment:
If the client sends one byte, and then just sits with the connection open without sending any more bytes, will this busy-loop?

Reply:
No, I don't think so. We don't seem to call `set_nonblocking` or anything like that in this module.

Reply:
On a second thought, I think you might be right! I'll look into this tomorrow.

Reply:
This code (both before and after) will not work correctly if the first read returns 1 to 3 bytes, because we'll throw away the bytes from the first read.

If this is about to be rewritten, then maybe it's not worth fixing now, but adding the `continue` doesn't seem right (as it will never result in correct behavior). Maybe it would be better to just exit with an error instead of silently doing the wrong thing...

Reply:
> This code (both before and after) will not work correctly if the first read returns 1 to 3 bytes, because we'll throw away the bytes from the first read.

That's not how `peek` works, though. https://docs.rs/tokio/1.9.0/tokio/net/struct.TcpStream.html#method.peek

Reply:
You're right, sorry! :)